File: leetcode/421-maximum-xor-of-two-numbers-in-an-array.py
class Solution(object):
    def findMaximumXOR(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        ans = 0
        for i in range(30, -1, -1):
            ans <<= 1
            prefix = {n>>i: 0 for n in nums}
            ans |= any(ans^1^p in prefix for p in prefix)
        return ans

File: leetcode/421-maximum-xor-of-two-numbers-in-an-array.cpp
class Solution
{
 public:
  int findMaximumXOR(vector<int>& nums)
  {
    int ans = 0;

    // If you move the following inside the loop, the memory for
    // previous iteration will not freed, and you will get a MLE.
    unordered_set<int> prefix;

    for (int i = 30; i >= 0; --i) {
      prefix.clear();

      for (int n : nums)
        prefix.insert(n>>i);

      ans <<= 1;
      for (int p : prefix)
        if (prefix.find(ans^1^p) != prefix.end()) {
          ans ^= 1;
          break;
        }
    }
    return ans;
  }
};
